Source/Core/VideoCommon/XFStructs.cpp

```diff
@@ -300,3 +301,431 @@ void PreprocessIndexedXF(u32 val, int refarray)
+                          fmt::format("Set projection[3] {}", Common::BitCast<float>(value)));
+  case XFMEM_SETPROJECTION + 4:
+    return std::make_pair(RegName(XFMEM_SETPROJECTION + 4),
+                          fmt::format("Set projection[4] {}", Common::BitCast<float>(value)));
+  case XFMEM_SETPROJECTION + 5:
+    return std::make_pair(RegName(XFMEM_SETPROJECTION + 5),
+                          fmt::format("Set projection[5] {}", Common::BitCast<float>(value)));
+  case XFMEM_SETPROJECTION + 6:
+  {
+    std::string desc;
+    if (value == GX_PERSPECTIVE)
+      desc = "Set projection type to perspective";
+    else if (value == GX_ORTHOGRAPHIC)
+      desc = "Set projection type to orthographic";
+    else
+      desc = "Set projection type to invalid value";
+    return std::make_pair(RegName(XFMEM_SETPROJECTION + 6), desc);
+  }
+
+  case XFMEM_SETNUMTEXGENS:  // GXSetNumTexGens
+    return std::make_pair(RegName(XFMEM_SETNUMTEXGENS),
+                          fmt::format("GXSetNumTexGens {}", value & 15));
+
+  case XFMEM_SETTEXMTXINFO:
+  case XFMEM_SETTEXMTXINFO + 1:
+  case XFMEM_SETTEXMTXINFO + 2:
+  case XFMEM_SETTEXMTXINFO + 3:
+  case XFMEM_SETTEXMTXINFO + 4:
+  case XFMEM_SETTEXMTXINFO + 5:
+  case XFMEM_SETTEXMTXINFO + 6:
+  case XFMEM_SETTEXMTXINFO + 7:
+    return std::make_pair(
+        fmt::format("XFMEM_SETTEXMTXINFO Matrix {}", address - XFMEM_SETTEXMTXINFO),
+        fmt::format("Matrix {}: {}", address - XFMEM_SETTEXMTXINFO, TexMtxInfo{.hex = value}));
+
+  case XFMEM_SETPOSTMTXINFO:
+  case XFMEM_SETPOSTMTXINFO + 1:
+  case XFMEM_SETPOSTMTXINFO + 2:
+  case XFMEM_SETPOSTMTXINFO + 3:
+  case XFMEM_SETPOSTMTXINFO + 4:
+  case XFMEM_SETPOSTMTXINFO + 5:
+  case XFMEM_SETPOSTMTXINFO + 6:
+  case XFMEM_SETPOSTMTXINFO + 7:
+    return std::make_pair(
+        fmt::format("XFMEM_SETPOSTMTXINFO Matrix {}", address - XFMEM_SETPOSTMTXINFO),
+        fmt::format("Matrix {}: {}", address - XFMEM_SETPOSTMTXINFO, TexMtxInfo{.hex = value}));
+
+  // --------------
+  // Unknown Regs
+  // --------------
+
+  // Maybe these are for Normals?
+  case 0x1048:  // xfmem.texcoords[0].nrmmtxinfo.hex = data; break; ??
+  case 0x1049:
+  case 0x104a:
+  case 0x104b:
+  case 0x104c:
+  case 0x104d:
+  case 0x104e:
+  case 0x104f:
+    return std::make_pair(
+        fmt::format("Possible Normal Mtx XF reg?: {:x}={:x}", address, value),
+        "Maybe these are for Normals? xfmem.texcoords[0].nrmmtxinfo.hex = data; break; ??");
+    break;
+
+  case 0x1013:
+  case 0x1014:
+  case 0x1015:
+  case 0x1016:
+  case 0x1017:
+
+  default:
+    return std::make_pair(fmt::format("Unknown XF Reg: {:x}={:x}", address, value), "");
+  }
+#undef RegName
+#undef DescriptionlessReg
+}
+
+std::pair<std::string, std::string> GetXFTransferInfo(const u8* data)
+{
+  const u32 cmd = Common::swap32(data);
+  data += 4;
+  u32 base_address = cmd & 0xFFFF;
+  const u32 transfer_size = ((cmd >> 16) & 15) + 1;
+
+  if (base_address > 0x1058)
```

Comment:
I'm not really good with this low level stuff.  But where do these hex values come from?  Maybe a comment unless it's just normal known information

Reply:
Short answer: they come from copying code in `XFRegWritten` and `LoadXFReg`.  Longer answer: it's `XFMEM_SETPOSTMTXINFO + 7` + 1, or one after the last register.  Probably there isn't actually a reason to treat it specially (since there's already a warning when writing to unknown registers), but there probably should be constants for that (`XFMEM_REGISTERS_END` maybe) and the start of the registers too.